src/hotspot/share/cds/filemap.cpp

```diff
@@ -1904,15 +1910,23 @@ bool FileMapInfo::relocate_pointers_in_core_regions(intx addr_delta) {
   if (bitmap_base == nullptr) {
     return false; // OOM, or CRC check failure
   } else {
-    size_t ptrmap_size_in_bits = header()->ptrmap_size_in_bits();
-    log_debug(cds, reloc)("mapped relocation bitmap @ " INTPTR_FORMAT " (" SIZE_FORMAT " bits)",
-                          p2i(bitmap_base), ptrmap_size_in_bits);
+    BitMapView rw_ptrmap = region_at(MetaspaceShared::rw)->ptrmap_view();
+    char* ro_bitmap_base = bitmap_base + rw_ptrmap.size_in_bytes();
+    BitMapView ro_ptrmap = region_at(MetaspaceShared::ro)->ptrmap_view();
 
-    BitMapView ptrmap((BitMap::bm_word_t*)bitmap_base, ptrmap_size_in_bits);
+    log_debug(cds, reloc)("mapped relocation rw bitmap @ " INTPTR_FORMAT " (" SIZE_FORMAT " bits)",
+                          p2i(bitmap_base), rw_ptrmap.size());
+    log_debug(cds, reloc)("mapped relocation ro bitmap @ " INTPTR_FORMAT " (" SIZE_FORMAT " bits)",
+                          p2i(ro_bitmap_base), ro_ptrmap.size());
```

Comment:
As we are using `BitMapView` in various cases (ptrmaps for RO and RW regions, and ptrmap/oopmap for HP region), I think we should simplify and unify the code.

The address of `ro_bitmap_base` and `bitmap_base` are internal to `FileMapRegion::bitmap_view()`, so we should move the logging inside that function.

I would suggest doing this:

```
BitMapView FileMapRegion::bitmap_view(bool is_oopmap, const char* name) {
  char* bitmap_base = FileMapInfo::current_info()->map_bitmap_region();
  bitmap_base += is_oopmap ? _oopmap_offset : _ptrmap_offset;
  size_t size_in_bits = is_oopmap ? _oopmap_size_in_bits : _ptrmap_size_in_bits;
  log_debug(cds, reloc)("mapped %s relocation %smap @ " INTPTR_FORMAT " (" SIZE_FORMAT " bits)",
                        name, is_oopmap ? "oop" : "ptr",
                        p2i(bitmap_base), size_in_bits);
  return BitMapView((BitMap::bm_word_t*)(bitmap_base), size_in_bits);
}
```

It's a shame that we have to pass in the name of the region, as a `FileMapRegion` doesn't know what region it is. I think we can remove these two functions:

```
BitMapView FileMapRegion::oopmap_view() {
  return bitmap_view(true);
}

BitMapView FileMapRegion::ptrmap_view() {
  assert(has_ptrmap(), "must be");
  return bitmap_view(false);
}
``` 

and add two corresponding functions in `FileMapInfo` (which knows the names of its regions):

```
BitMapView FileMapInfo::oopmap_view(int region_index) {
  return region_at(region_index)->bitmap_view(/*is_oopmap*/true, region_name(region_index));
}

BitMapView FileMapInfo::ptrmap_view(int region_index) {
  return region_at(region_index)->bitmap_view(/*is_oopmap*/false, region_name(region_index));
}
```

The assert can be moved into FileMapRegion::bitmap_view:

```
  if (is_oopmap) {
    assert(has_oopmap(), "sanity");
  } else {
    assert(has_ptrmap(), "sanity");
   }
```

There are a few places that we construct the BitMapView by hand. E.g.,

```
void ArchiveHeapLoader::patch_embedded_pointers(FileMapInfo* info,
                                                MemRegion region, address oopmap,
                                                size_t oopmap_size_in_bits) {
  BitMapView bm((BitMap::bm_word_t*)oopmap, oopmap_size_in_bits);
```

The `oopmap` and `oopmap_size_in_bits` parameters should be removed and this function should call

```
BitMapView vm = info->oopmap(MetaspaceShared::hp);
```

There's a similar issue with `ArchiveHeapLoader::load_heap_region_impl()`.